data/core/docview.lua

```diff
@@ -364,9 +364,17 @@ end
 function DocView:draw_line_text(line, x, y)
   local default_font = self:get_font()
   local tx, ty = x, y + self:get_line_text_y_offset()
-  for _, type, text in self.doc.highlighter:each_token(line) do
+  local last_token = nil
+  local tokens = self.doc.highlighter:get_line(line).tokens
+  local tokens_count = #tokens
+  if string.sub(tokens[tokens_count], -1) == "\n" then
+    last_token = tokens_count - 1
+  end
+  for tidx, type, text in self.doc.highlighter:each_token(line) do
     local color = style.syntax[type]
     local font = style.syntax_fonts[type] or default_font
+    -- do not render newline, fixes issue #1164
+    if tidx == last_token then text = text:sub(1, -2) end
```

Comment:
I wonder if there's some important overhead with extremely long tokens.

Reply:
here some test using your bench function:

```lua
---Generate really long text
local hello = string.rep("test", 1000000)

---Benchmark string.sub
common.bench("test", function() hello:sub(1, -2) end)

---Results when calling it more than once:
-- *** test             :   14.058ms  84.35%
-- *** test             :    1.825ms  10.95%
-- *** test             :    2.135ms  12.81%
```

Reply:
I mean with *long* strings. As it'd have to basically duplicate it, I wonder if that might create a bottleneck.

(Also, I didn't make that function, it's there since the very beginning.)

Reply:
Ah I missed the `string.rep`, my bad.